runtime/src/main/java/org/corfudb/runtime/LiteRoutingQueueListener.java

```diff
@@ -72,118 +74,143 @@ public LiteRoutingQueueListener(CorfuStore corfuStore, String sourceSiteId, Stri
+                .findFirst().get();
+
         List<Table.CorfuQueueRecord> allEntries = new ArrayList<>(entries.size());
-        Queue.ReplicationType currentType = Queue.ReplicationType.LAST_FULL_SYNC_ENTRY;
+        List<Queue.RoutingTableEntryMsg> allRQMsgs = new ArrayList<>();
+
+        // Capture the first message in this batch and compute the current replication type.  It will later be used
+        // for validation that all entries in a transaction belong to the same replication type.
+        // For snapshot sync, the first message will be CLEAR so get the next message, if any.
+        // For log entry sync, get the first message.
+        CorfuStreamEntry firstUpdateEntry = null;
+        if (entries.get(0).getOperation() == CorfuStreamEntry.OperationType.CLEAR  && entries.size() > 1) {
+            firstUpdateEntry = entries.get(1);
+        } else if (entries.get(0).getOperation() == CorfuStreamEntry.OperationType.UPDATE) {
+            firstUpdateEntry = entries.get(0);
+        }
+
+        if (firstUpdateEntry != null) {
+            Queue.RoutingTableEntryMsg firstMsg = (Queue.RoutingTableEntryMsg) firstUpdateEntry.getPayload();
+            currentReplicationType = firstMsg.getReplicationType();
+        }
+
         for (CorfuStreamEntry entry : entries) {
-            if (!entry.getOperation().equals(CorfuStreamEntry.OperationType.UPDATE)) {
+            // The Source always 'adds' replicated data to the queue.  So ignore all operations where type != UPDATE.
+            // Non-update entries are written by LR (CLEAR) and this listener itself (DELETE after successful
+            // processing).
+            if (entry.getOperation() != CorfuStreamEntry.OperationType.UPDATE) {
                 continue;
             }
             Queue.CorfuGuidMsg key = (Queue.CorfuGuidMsg) entry.getKey();
             Queue.RoutingTableEntryMsg msg = (Queue.RoutingTableEntryMsg) entry.getPayload();
+
+            if (msg.getReplicationType() != currentReplicationType) {
+                throw new LogReplicationClientException("Not expecting mixed event types "+ currentReplicationType +" vs "
+                    + msg.getReplicationType());
+            }
+
+            allRQMsgs.add(msg);
             Queue.CorfuQueueMetadataMsg metadataMsg = (Queue.CorfuQueueMetadataMsg) entry.getMetadata();
             allEntries.add(new Table.CorfuQueueRecord(key, metadataMsg, msg));
-            if (currentType.equals(Queue.ReplicationType.LAST_FULL_SYNC_ENTRY)) {
-                currentType = msg.getReplicationType();
-            } else if (!currentType.equals(msg.getReplicationType())) {
-                throw new LogReplicationClientException("Not expecting mixed event types "+ currentType +" vs "
-                    +msg.getReplicationType());
-            }
         }
         if (allEntries.isEmpty()) {
             return;
         }
```

Comment:
Will the `DELETE` type entries written by this listener itself fall into this check? maybe we can add a log msg to indicate that or even explicitly drop out DELETE case?

Reply:
yes, we filter out any entries with type != CLEAR on L104